Expensify/react-native
# Substr measurements expensify

# Upstream PR Link
https://github.com/facebook/react-native/pull/37397


## Summary



I would like to apply styling to a substring of text, such as a background color, padding, and rounded corners. This functionality could be used for:

- **Username mention highlighting** (similar to Slack): Happy birthday `@username`
- **Inline code block** (similar to GitHub): Lorem Ipsum `is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen` book. 
<img width="443" alt="Screen Shot 2023-05-16 at 14 42 00" src="https://github.com/facebook/react-native/assets/4882133/393416d9-2a28-4653-9130-29d8c2e0dfdc">

The problem is that the Text element in RN behaves like a block element rather than an inline element when it is multiline. Additionally, it does not allow you to have borders or rounded corners. This makes it unsuitable for implementing a multiline inline code block.

Adding a new prop called textLayoutRegions which accepts an array of regions (substring positions) could be used to solve the problem. Passing this prop will calculate the regions [[start, end], [start, end] ...] position and pass the response into onTextLayout callback. Knowing exact substring position and dimensions will allow implementing custom styled View behind the text.

```js
<View>
  {regions.map((region, index) => (
    <View style={{width: region.width, height: region.height, top: region.y, left: region.x, backgroundColor: 'red', position: 'absolute'}} />
  ))}

  <Text style={{lineHeight: 20}} onTextLayout={event => setRegions(event.nativeEvent.regions)} textLayoutRegions={[[12, 242]]}>
    Lorem Ipsum  is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen  book.
  </Text>
</View>

// onTextLayout event
{
  lines: [...],
  regions: [
  {
    "height": 20,
    "line": 0,
    "region": 0,
    "text": " is simply dummy text of the printing and typesetting ",
    "width": 325.71429443359375,
    "x": 85.42857360839844,
    "y": 0
  },
  {
    "height": 20,
    "line": 1,
    "region": 0,
    "text": "industry. Lorem Ipsum has been the industry's standard dummy ",
    "width": 393.71429443359375,
    "x": 0,
    "y": 20
  },
  {
    "height": 20,
    "line": 2,
    "region": 0,
    "text": "text ever since the 1500s, when an unknown printer took a galley ",
    "width": 400.28570556640625,
    "x": 0,
    "y": 40
  },
  {
    "height": 20,
    "line": 3,
    "region": 0,
    "text": "of type and scrambled it to make a type specimen  ",
    "width": 313.1428527832031,
    "x": 0,
    "y": 60
  }
]
}
```

## Changelog



[GENERAL] [ADDED] - New `textLayoutRegions` prop for Text component that accepts an array of substring regions. Each region's layout position will be appended into `onTextLayout` event callback.
[GENERAL] [ADDED] - New `regions` props for `onTextLayout` event callback.


## Test Plan



- Run RNTester app
- Go to APIs → Layout Events page
- Ensure substring has a red background color around selected region
